### ISSUE TITLE
HttpClient throws an exception when the remote closes the Keep-alive socket

### ISSUE THREAD

**Issue by ShieldHero** [nim-lang/Nim#12299]
This line https://github.com/nim-lang/Nim/blob/version-1-0/lib/pure/httpclient.nim#L783 makes `requestAux` skip making a new connection, which throws an exception in case the `client.connected` is true when the remote server already closed the connection.

### Possible Solution

Nim should handle this case, so far the only workaround is to wrap the calling block in a try/except and retry, the second try will always succeed since `newConnection` will be called properly.

this line https://github.com/nim-lang/Nim/blob/version-1-0/lib/pure/httpclient.nim#L796 is redundant, the `close` proc already handles setting connected to false.

### Additional Information

```
$ nim -v
Nim Compiler Version 1.0.99 [Windows: i386]
Compiled at 2019-09-26
Copyright (c) 2006-2019 by Andreas Rumpf

git hash: 657e09e79deabe6304ead7759a4fdcfeb5fd022b
active boot switches: -d:release
```

**Comment by ShieldHero** [comment]
this might be related to https://github.com/nim-lang/Nim/pull/7530

**Comment by disruptek** [comment]
**TL;DR** `Keep-Alive` means `close only on my terms`, it doesn't mean `Reconnect`.

The right solution is not to reconnect, because that means going through a new process on the server which may fail.  For example, if my requests are signed, then after a reconnection my signature may fail, in which case the client will yield a 4xx error to the code above.  It's easy to imagine other scenarios in which I don't want to reconnect to the server if I lose my connection, and that's why `Keep-Alive` doesn't specify reconnection.  I'd rather an error state that my client can inspect to figure out that the connection was merely torn down prior to a subsequent request.

**Comment by ShieldHero** [comment]
@disruptek The problem is that I didn't see a better way to check for the error elegantly other than try/except. And after checking other people's code, they rarely handle this case, which leads to random crashes. I think most people assume that it will never fail in that particular way. The exception message isn't really that helpful either. 

A new functions that will retry automatically or a cleaner way to check for the error is very much appreciated.

More info: My use case is a discord bot, most of the time it works fine as long as I keep the http requests going, but any idle time will cause a crash, it happens randomly too.

**Comment by dom96** [comment]
@ShieldHero does Discord use websockets? If so, shouldn't you be sending PINGs to the server to prevent that from happening?

**Comment by ShieldHero** [comment]
@dom96 Yes discord does use websockets, but AFAIK HTTP requests are handled separately. Basically the websocket sends heartbeats to keep that connection alive, but it doesn't affect the HTTP client in the slightest. Each HTTP request can be a new connection. I'm not aware of anyway to keep HTTP connection alive from the server side, even then, the discord bot can be idle for any amount of time and the server can close the HTTP connection without notification.

**Comment by dom96** [comment]
websockets works over http, if your Discord bot is idle it can still be sending PINGs which should ensure that the connection is kept alive. But maybe Discord works differently, I don't know, this is just my expectation.
